src/main/java/com/android/volley/toolbox/DiskBasedAsyncCache.java

```diff
@@ -0,0 +1,145 @@
+
+    /** High water mark percentage for the cache */
+    @VisibleForTesting static final float HYSTERESIS_FACTOR = 0.9f;
+
+    /** Magic number for current version of cache file format. */
+    private static final int CACHE_MAGIC = 0x20150306;
+
+    /**
+     * Constructs an instance of the DiskBasedAsyncCache at the specified directory.
+     *
+     * @param rootDirectory The root directory of the cache.
+     * @param maxCacheSizeInBytes The maximum size of the cache in bytes. Note that the cache may
+     *     briefly exceed this size on disk when writing a new entry that pushes it over the limit
+     *     until the ensuing pruning completes.
+     */
+    public DiskBasedAsyncCache(final File rootDirectory, int maxCacheSizeInBytes) {
+        mRootDirectorySupplier =
+                new DiskBasedCacheUtility.FileSupplier() {
+                    @Override
+                    public File get() {
+                        return rootDirectory;
+                    }
+                };
+        mMaxCacheSizeInBytes = maxCacheSizeInBytes;
+    }
+
+    /** Clears the cache. Deletes all cached files from disk. */
+    @Override
+    public synchronized void clear() {
+        // TODO (sphill99): Implement
+    }
+
+    /** Returns the cache entry with the specified key if it exists, null otherwise. */
+    @Override
+    public void get(String key, OnGetCompleteCallback callback) {
+        final OnGetCompleteCallback cb = callback;
+        final CacheHeader entry = mEntries.get(key);
+        // if the entry does not exist, return.
+        if (entry == null) {
+            cb.onGetComplete(null);
+            return;
+        }
+        File file = getFileForKey(key);
+        Path path = Paths.get(file.getPath());
+        try {
+            AsynchronousFileChannel afc =
+                    AsynchronousFileChannel.open(path, StandardOpenOption.READ);
+            ByteBuffer buffer = ByteBuffer.allocate((int) file.length());
+            afc.read(
+                    /* destination */ buffer,
+                    /* position */ 0,
+                    /* attachment */ buffer,
+                    new CompletionHandler<Integer, ByteBuffer>() {
+                        @Override
+                        public void completed(Integer result, ByteBuffer attachment) {
+                            if (attachment.hasArray()) {
+                                int offset = attachment.arrayOffset();
+                                byte[] data = attachment.array();
+                                cb.onGetComplete(entry.toCacheEntry(data));
+                            }
+                        }
+
+                        @Override
+                        public void failed(Throwable exc, ByteBuffer attachment) {
+                            cb.onGetComplete(null);
+                        }
+                    });
+        } catch (IOException e) {
+            cb.onGetComplete(null);
```

Comment:
log exception

Reply:
fixed

Reply:
better, but see above note regarding how to handle errors

Reply:
Line 107 is still potentially a second call to the callback

Reply:
Line 107 is still potentially a second invocation of callback.

See comment above and this example of another way to structure this:

https://github.com/google/volley/blob/master/src/main/java/com/android/volley/toolbox/DiskBasedCache.java#L148